contents/data/nosql/NoSQL.md

```diff
@@ -0,0 +1,437 @@
+
+
+![](documentDatabaseStructure.png)
+
+In contrast to tables and table entries in SQL databases, document dtabases comprise of collections and documents.
+A database comprises of multiple collections, which in turn consists of multiple documents.
+
+In a simplified e-commerce website for example, you may have the following collections: { id="main-example" }
+- _customers_ - storing the account details of customers, their purchase histories, etc.
+- _items_ - a collection of all items available for purchase (which are documents)
+- _admin_ - a collection storing admin account details
+- **...**
+
+In this case, the _items_ collection which contains the catalogue of purchase items may be structured like so:
+
+<div id="schemaless-example">
+
+```js
+[
+  {
+    type: "book",
+    title: "about pandas",
+    price: 20,
+    popularity: 9.7,
+    author: "panda1",
+    ...
+  },
+  {
+    type: "grocery",
+    name: "cheese",
+    brand: "panda",
+  },
+  ...
+]
+```
+
+</div>
+
+<br>
+<box type="info" light>
+Note that while many NoSQL databases provide a JSON interface to interact with the data,
+the underlying storage implementation may be different for performance reasons.<br>
+For example, MongoDB stores documents in BSON (json encoded in binary). 
+</box>
+
+<br>
+
+## Key Characteristics of NoSQL Databases
+
+### 1. Powerful and Simple CRUD Operations
+
+Interaction with most document databases is achieved in a simple and intuitive _object oriented manner_,
+and JSON-like queries.
+
+Let's get back to the above [example](#main-example) of an e-commerce website. To register a new user account,
+an example query made in mongoDB would be like so:
+
+```js
+db.customers.insertOne({
+  username: "panda",
+  password: hashedPassword,
+  email: "panda@pandas.com"
+})
+```
+
+Like SQL databases, queries in NoSQL databases often also support more specific and powerful variants, and can
+be even more succinct in some cases.
+
+For example, to filter through items with a price of less than 30, and a popularity of more than 8,
+thereafter sorting them by their price, you would make a query in MongoDB like so, intuitively matching
+the structure of a typical item in the _items_ collection.
+
+```js
+db.items.find({
+  price: {
+    $lt: 30
+  },
+  popularity: {
+    $gt: 8
+  }
+}).sort({
+  price: 1 // Here 1 means ascending order
+})
+```
+
+<br>
+
+### 2. Schema-Less Data
+
+Another key characteristic of most NoSQL databases is that they are **schema-less**.
+In document databases for example, this means that each individual document has no restriction on what keys it must have,
+the number of keys, the type of values and so on.
+
+<box type="info" light>
+Note the missing fields for the second item in the <em>items</em> collection
+<popover header="Example items collection of an e-commerce website">earlier,<template slot="content"><include src="NoSQL.md#schemaless-example" /></template></popover>
+which are intentionally omitted.
```

Comment:
How come the text in this popover appear smaller than regular text? 🤔

Reply:
Changed this to a modal; Popovers are hardcoded to be `< 276px` in bootstrap apparently.  ( for mobile-friendliness )

Reply:
Actually, I was referring to the font size. Popovers use smaller font size?